### DISCUSSION TITLE
Help with correctly configuring the camera

### DISCUSSION THREAD

**Discussion by MikeLerwill99** [raspberrypi/picamera2#567] [Q&A]
If I use the following configuration with a camera 3 wide

picam2.configure(picam2.create_video_configuration(main={"size": (1440, 1440), "format": "RGB888" }, 
    lores={"size": (440, 440), "format": "YUV420"}, encode = "lores"))

picam2.set_controls ({"ScalerCrop":(1400,450,1800,1800)})

The stream from the lores configuration is as I want it and a captured frame from the main configuration 
is also correct but is twice the size that I want and I am having to resize it to 720x720.

If however I use the following instead

picam2.configure(picam2.create_video_configuration(main={"size": (720, 720), "format": "RGB888" }, 
    lores={"size": (440, 440), "format": "YUV420"}, encode = "lores"))

picam2.set_controls ({"ScalerCrop":(1400,450,1800,1800)})

While the sizes are now as I want them the cropped area is not correct.

This would appear to be because of the sensor image format, in the first case it is reporting  

Selected sensor format: 2304x1296-SBGGR10_1X10 - Selected unicam format: 2304x1296-pBAA

while in the second case it is reporting

Selected sensor format: 1536x864-SBGGR10_1X10 - Selected unicam format: 1536x864-pBAA

Is there a way I can force it to use the larger sensor format while still setting the size to 720x720

Thanks

**Comment by sandyol55** [comment]
Setting an additional 'raw' configuration in the mode you want to force the system to use is one way to do it.

eg something like 
picam2.configure(picam2.create_video_configuration(main={"size": (720, 720), "format": "RGB888" },
lores={"size": (440, 440), "format": "YUV420"}, raw={"size": (2304,1296)},encode = "lores"))

or more generally to pick up the available sensor modes with picam2.sensor_modes and then apply the appropriate one with raw=picam2.sensor_modes[x] as per Section 4.2.2 of the manual.
[https://datasheets.raspberrypi.com/camera/picamera2-manual.pdf]

**Reply by MikeLerwill99** [reply]
Thanks,

Setting the raw size as you suggest works for me, I tried doing it using sensor_modes[2] as per your other suggestion but could not find a correct syntax for it

using 
picam2.configure(picam2.create_video_configuration(main={"size": (720,720), "format": "RGB888" }, 
    lores={"size": (440,440), "format": "YUV420"}, raw = picam2.sensor_modes[2],  encode = "lores"))

gives me a error
  File "/usr/lib/python3/dist-packages/picamera2/picamera2.py", line 597, in _make_initial_stream_config
    raise ValueError(f"Bad key '{key}': valid stream configuration keys are {valid}")
ValueError: Bad key 'unpacked': valid stream configuration keys are ('format', 'size')

and I'm not sure what its complaining about

regards

Mike Lerwill

**Reply by dan-1d** [reply]
FYI, I also have this same problem. Likely an argument parsing error

**Reply by sandyol55** [reply]
> FYI, I also have this same problem. Likely an argument parsing error

Think this has been fixed with [PR](url)https://github.com/raspberrypi/picamera2/pull/637 so should be in the next branch now?

**Comment by sandyol55** [comment]
Yea, I see the same. Hopefully someone more knowledgeable will advise how-to correctly, but a workaround that allows setting some of the basic configurations without knowing the specific sensor sizes is as below.

```
picam2 = Picamera2()

full_res=picam2.sensor_resolution

half_res=tuple([dim // 2 for dim in picam2.sensor_resolution])

#configure full format highest res mode RGB 
still_config=picam2.create_still_configuration(main={"size":(720,720),"format":"RGB888"}, raw={"size":full_res})

#configure full format half resolution binned mode BGR
still_config=picam2.create_still_configuration(main={"size":(720,720),"format":"BGR888"}, raw={"size":half_res})

picam2.configure(still_config)

picam2.start()
```
HTH

Sandy

**Comment by davidplowman** [comment]
That all sounds reasonable to me - specifying the raw stream resolution is the way you force the camera mode that you want. The notion of a "full resolution mode" and a "2x2 binned mode" is generally true - except when it isn't (Global Shutter cam, for example)! In the end, the only way to be sure you have the best mode for any particular application, is to look through them and decide (or program your application to decide) for itself.

For this reason there's the `sensor_modes` property. Be sure to fetch this as soon as you open the camera because it will reconfigure the sensor multiple times to get all the information, so you can't do it once you've started doing something. So:
```
from picamera2 import Picamera2
picam2 = Picamera2()
sensor_modes = picam2.sensor_modes
```
That gives you a list of all the camera modes that truly exist, as well as information about them, such as resolution, max framerate, field of view, so in theory you can make all those trade-offs for yourself.
